In [1]:
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline

In [2]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler, MinMaxScaler, LabelEncoder
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import RandomForestRegressor
from datetime import datetime
import jenkspy

In [3]:
Walmart=pd.read_csv("Walmart.csv")
Walmart.head(10)

Store        Date  Weekly_Sales  Holiday_Flag  Temperature  Fuel_Price  \
0      1  05-02-2010    1643690.90             0        42.31       2.572   
1      1  12-02-2010    1641957.44             1        38.51       2.548   
2      1  19-02-2010    1611968.17             0        39.93       2.514   
3      1  26-02-2010    1409727.59             0        46.63       2.561   
4      1  05-03-2010    1554806.68             0        46.50       2.625   
5      1  12-03-2010    1439541.59             0        57.79       2.667   
6      1  19-03-2010    1472515.79             0        54.58       2.720   
7      1  26-03-2010    1404429.92             0        51.45       2.732   
8      1  02-04-2010    1594968.28             0        62.27       2.719   
9      1  09-04-2010    1545418.53             0        65.86       2.770   

          CPI  Unemployment  
0  211.096358         8.106  
1  211.242170         8.106  
2  211.289143         8.106  
3  211.319643         8.106  
4  211.350143         8.106  
5  211.380643         8.106  
6  211.215635         8.106  
7  211.018042         8.106  
8  210.820450         7.808  
9  210.622857         7.808

In [4]:
Walmart.isnull().sum()

Store           0
Date            0
Weekly_Sales    0
Holiday_Flag    0
Temperature     0
Fuel_Price      0
CPI             0
Unemployment    0
dtype: int64

In [5]:
Walmart3=Walmart.rename(columns={'Weekly_Sales': 'Sales'})

In [6]:
Walmart3.drop(columns=["Date"], inplace=True)

In [7]:
from sklearn import preprocessing
le = preprocessing.LabelEncoder()

In [8]:
Walmart3['Sales'] = Walmart3['Sales'].astype(str)

In [9]:
Walmart3['label_sales']=''
for index, row in Walmart3.iterrows():
    if (row['Sales']<='100000'):
            Walmart3.loc[index, 'label_sales']=0
    elif(row['Sales']>'100000')& (row['Sales']<='500000'):
                Walmart3.loc[index, 'label_sales']=1
    elif(row['Sales']>'500000')& (row['Sales']<='1000000'):
                Walmart3.loc[index, 'label_sales']=2
    elif(row['Sales']>'1000000')& (row['Sales']<='3000000'):
                Walmart3.loc[index, 'label_sales']=3
    elif(row['Sales']>'300000')& (row['Sales']<='500000'):
                Walmart3.loc[index, 'label_sales']=4
    elif(row['Sales']>'500000')& (row['Sales']<='800000'):
                Walmart3.loc[index, 'label_sales']=5
    elif(row['Sales']>'800000')& (row['Sales']<='1000000'):
                Walmart3.loc[index, 'label_sales']=6
    elif(row['Sales']>'10000000')& (row['Sales']<='30000000'):
                Walmart3.loc[index, 'label_sales']=7
    elif(row['Sales']>'30000000')& (row['Sales']<='50000000'):
                Walmart3.loc[index, 'label_sales']=8
    elif(row['Sales']>'50000000')& (row['Sales']<='80000000'):
                Walmart3.loc[index, 'label_sales']=9
    elif(row['Sales']>'80000000')& (row['Sales']<='1000000000'):
                Walmart3.loc[index, 'label_sales']=10
    elif(row['Sales']>'100000000')& (row['Sales']<='300000000'):
                Walmart3.loc[index, 'label_sales']=11
    elif(row['Sales']>'300000000')& (row['Sales']<='5000000000'):
                Walmart3.loc[index, 'label_sales']=12
    else: 
                Walmart3.loc[index, 'label_sales']=0

In [10]:
Walmart3['label_sales'].value_counts()

1    4237
5    1344
0     854
Name: label_sales, dtype: int64

In [11]:
y = Walmart3['label_sales'].astype('int')
X = Walmart3.drop(['label_sales', 'Sales'], axis = 1)

In [12]:
y

0       1
1       1
2       1
3       1
4       1
       ..
6430    5
6431    5
6432    5
6433    5
6434    5
Name: label_sales, Length: 6435, dtype: int32

In [13]:
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)

In [14]:
y_train

6246    1
3855    1
809     1
1807    1
5954    5
       ..
3772    1
5191    5
5226    5
5390    1
860     5
Name: label_sales, Length: 4826, dtype: int32

In [15]:
scaler = StandardScaler().fit(X_train)
X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)
X_train_scaled

array([[ 1.60047202, -0.26871207, -1.76541795, -0.17397815, -1.06160558,
        -1.02955636],
       [ 0.29160115, -0.26871207,  0.31009121,  1.83998324, -0.73557291,
         0.12674183],
       [-1.32523933,  3.72145546,  0.12994244, -0.27456688,  1.23774926,
        -0.77646564],
       ...,
       [ 1.06152519, -0.26871207,  1.42842437,  0.70507992,  1.09122181,
         0.09356715],
       [ 1.1385176 ,  3.72145546, -0.85436439,  0.14528175, -1.05719358,
         2.61537805],
       [-1.24824693, -0.26871207, -1.7963471 , -1.77465104,  0.45957052,
         0.54142535]])

In [16]:
classifier = LogisticRegression()
classifier

LogisticRegression()

In [17]:
classifier.fit(X_train, y_train)

C:\Users\Stephanie\Python\lib\site-packages\sklearn\linear_model\_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [18]:
y_predict = classifier.predict(X_test)
y_predict

array([1, 1, 1, ..., 1, 1, 5])

In [19]:
classifier.score(X_train, y_train)

0.6469125569830088

In [20]:
classifier.score(X_test, y_test)

0.6712243629583592

In [21]:
y_pred = pd.DataFrame(classifier.predict(X_test), index=X_test.index, columns=['Sales Projection'])
y_pred_df = y_pred.merge(y_test, how='outer', left_index=True, right_index=True)
y_pred_df

Sales Projection  label_sales
2436                 1            1
3361                 1            1
233                  1            1
3667                 1            1
5011                 1            1
...                ...          ...
4272                 1            1
1094                 1            0
1121                 1            0
3737                 1            1
5579                 5            0

[1609 rows x 2 columns]

__________________________________________________________________________________________________

In [22]:
print(f"Training Data Score: {classifier.score(X_train, y_train)}")
print(f"Testing Data Score: {classifier.score(X_test, y_test)}")

Training Data Score: 0.6469125569830088
Testing Data Score: 0.6712243629583592


Logistic Regression is really only meant to be used for binary prediction - "yes or no", "1 or 2" -- not best option for predicting future sales projections which is why the accuracy is only about 65%.